From 83e6039abd4fc9e734b7b0a8ba81824627aa450b Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 14 Feb 2022 08:49:03 +0200
Subject: [PATCH] [SW-75064] gaudi2: irq: ack pci irqs

PCIE_ADDR_DEC_ERR irq is showing after hard reset, sometimes
followed by pci drain irq.
As agreed before, IRQs that are consequence of the reboot sequence
should be cleared in preboot.

Change-Id: If3055955f3669a19862d83e1bbdb821a0bc64ffc
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index f1f500a..4a0b3eb 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -49,6 +49,7 @@
 #include <psoc_arc1_cfg_masks.h>
 #include <psoc_uart_0_regs.h>
 #include <pcie_elbi_rr_mstr_if_rr_shrd_lbw_regs.h>
+#include <gic_regs.h>
 
 #include "hl.h"
 #include "common.h"
@@ -1302,3 +1303,26 @@
 {
 	ca53_cfg.arm_rst_control_reg = mmCPU_CA53_CFG_ARM_RST_CONTROL;
 }
+
+#define GAUDI2_IRQ_SIZE	512
+static void gaudi2_ack_gic_irq(void)
+{
+	u32 i;
+
+	for (i = 0; i < GAUDI2_IRQ_SIZE; i++)
+		regmap_writel(i, mmGIC_DISTRIBUTOR__6_GICD_CLRSPI_NSR);
+}
+
+static void gaudi2_ack_pci_axi_irq(void)
+{
+	/* Ack PCI addr decoder error */
+	regmap_writel(1, mmPCIE_WRAP_AXI_INTR);
+	/* Ack PCI drain indication */
+	regmap_writel(0, mmPCIE_WRAP_AXI_DRAIN_IND);
+}
+
+void preboot_ack_irq(void)
+{
+	gaudi2_ack_pci_axi_irq();
+	gaudi2_ack_gic_irq();
+}
